applications/CompressiblePotentialFlowApplication/custom_processes/define_3d_wake_process.cpp

```diff
@@ -0,0 +1,879 @@
+                                         ModelPart& rBodyModelPart,
+                                         ModelPart& rStlWakeModelPart,
+                                         const double Tolerance,
+                                         const Vector& rWakeNormal,
+                                         const Vector& rWakeDirection,
+                                         const bool SwitchWakeDirection,
+                                         const bool CountElementsNumber,
+                                         const bool WriteElementsIdsToFile,
+                                         const bool ShedWakeFromTrailingEdge,
+                                         const double SheddedWakeDistance,
+                                         const double SheddedWakeElementSize)
+    : Process(),
+      mrTrailingEdgeModelPart(rTrailingEdgeModelPart),
+      mrBodyModelPart(rBodyModelPart),
+      mrStlWakeModelPart(rStlWakeModelPart),
+      mTolerance(Tolerance),
+      mWakeNormal(rWakeNormal),
+      mWakeDirection(rWakeDirection),
+      mSwitchWakeDirection(SwitchWakeDirection),
+      mCountElementsNumber(CountElementsNumber),
+      mWriteElementsIdsToFile(WriteElementsIdsToFile),
+      mShedWakeFromTrailingEdge(ShedWakeFromTrailingEdge),
+      mSheddedWakeDistance(SheddedWakeDistance),
+      mSheddedWakeElementSize(SheddedWakeElementSize)
+{
+    KRATOS_ERROR_IF(mWakeNormal.size() != 3)
+        << "The mWakeNormal should be a vector with 3 components!"
+        << std::endl;
+}
+
+void Define3DWakeProcess::ExecuteInitialize()
+{
+    ModelPart& root_model_part = mrBodyModelPart.GetRootModelPart();
+    auto& r_nodes = root_model_part.Nodes();
+    VariableUtils().SetNonHistoricalVariable(UPPER_SURFACE, false, r_nodes);
+    VariableUtils().SetNonHistoricalVariable(LOWER_SURFACE, false, r_nodes);
+
+    InitializeTrailingEdgeSubModelpart();
+
+    InitializeWakeSubModelpart();
+
+    // Compute span direction as the cross product: mWakeNormal x mWakeDirection
+    MathUtils<double>::CrossProduct(mSpanDirection, mWakeNormal, mWakeDirection);
+
+    MarkTrailingEdgeNodesAndFindWingtipNodes();
+
+    ComputeWingLowerSurfaceNormals();
+
+    ComputeAndSaveLocalWakeNormal();
+
+    if(mShedWakeFromTrailingEdge){
+        KRATOS_INFO("Define3DWakeProcess") << " Shedding wake from the trailing edge" << std::endl;
+        ShedWakeSurfaceFromTheTrailingEdge();
+    }
+
+    MarkWakeElements();
+
+    RecomputeNodalDistancesToWakeOrWingLowerSurface();
+
+    MarkKuttaElements();
+
+    SaveLocalWakeNormalInElements();
+
+    AddWakeNodesToWakeModelPart();
+
+    if(mCountElementsNumber){
+        CountElementsNumber();
+    }
+
+    if(mWriteElementsIdsToFile){
+        WriteElementIdsToFile();
+    }
+}
+
+// This function initializes the variables and removes all of the elements and
+// nodes of the trailing edge submodelpart
+void Define3DWakeProcess::InitializeTrailingEdgeSubModelpart() const
```

Comment:
It doesn't modify it, does it?

Reply:
I also forgot to ask this :sweat_smile:

Reply:
OK i thought it was being modified since we are creating sub model parts and changing elements in the model part, but that's on the `root_model_part` (not the body, which remains the same).